### DISCUSSION TITLE
No mediaDevices in navigator interface

### DISCUSSION THREAD

**Discussion by stefanzvkvc** [keathley/wallaby#656] [Q&A]
Hi there,

I was looking for library that will help me to spin browser session and execute JavaScript in order to create WebRTC connection and so I came across your library.
Everything went as it should until the moment when I wanted to read `mediaDevices` object from `navigator` interface. Problem is that 
when I execute script:
  
    `let pc; 
    
    const offerOptions = {
      audio: true,
      video: true
    };

    switch (arguments[0]) {
      case 'offer': return createOffer();
    }

    function createOffer() {
      pc = new RTCPeerConnection(offerOptions);
      navigator.mediaDevices.getUser();
    }`
    
I receive `javascript error: Cannot read properties of undefined`

Do you have something on your mind how I can resolve this situation?

Thanks in advance! :)

My local environment is:

Elixir (1.12.3)
ChromeDriver: 96.0.4664.45
Chrome: 96.0.4664.110
platform: Mac OS X 10.15.7

**Comment by mhanberg** [comment]
Can you share your elixir code as well?

**Comment by stefanzvkvc** [comment]
Hey thanks for fast reply. Sure I can.
This is an example of the test.

```
defmodule Test do
  use ExUnit.Case, async: true
  use Wallaby.DSL

  setup do
    :ok = Ecto.Adapters.SQL.Sandbox.checkout(:otp_app)
    Ecto.Adapters.SQL.Sandbox.mode(:otp_app, {:shared, self()})
    .
    .
    .
    :ok
  end

  defp script() do
    """
    let pc;

    const offerOptions = {
      audio: true,
      video: true
    };

    switch (arguments[0]) {
      case 'offer': return createOffer();
    }

    function createOffer() {
      pc = new RTCPeerConnection(offerOptions);
      navigator.mediaDevices;
    }
    """
  end

  defp get_chrome_options() do
    %{
      chromeOptions: %{
        args: [
          "--headless",
          "--use-fake-device-for-media-stream",
          "--use-fake-ui-for-media-stream",
          "--user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36"
        ],
        prefs: %{
          "profile.default_content_setting_values.media_stream_mic": 1,
          "profile.default_content_setting_values.media_stream_camera": 1
        }
      }
    }
  end

  test "returns media devices object" do
    options = get_chrome_options()
    {:ok, session} = Wallaby.start_session(capabilities: options)

    execute_script(session, script(), ["offer"], fn value ->
      send(self(), {:callback, value})
    end)

    value =
      receive do
        {:callback, value} ->
          value
      after
        5000 -> nil
      end

    assert not is_nil(value)
  end
end
```

test_helper.exs
```
{:ok, _} = Application.ensure_all_started(:wallaby)
ExUnit.start()
```

**Reply by mhanberg** [reply]
There's a chance you need to have `window.navigator`. 

The script runs as a function body and has different scoping rules I think. 

You can see more here: https://www.w3.org/TR/webdriver2/#executing-script

**Comment by stefanzvkvc** [comment]
Thanks for the link I will definitely check it 👍 

I've also tried `window.navigator` but without success. Maybe this is a permission issue for Chrome?

I will try and check results with GeckoDriver.

**Comment by stefanzvkvc** [comment]
@mhanberg I've tried with GeckoDriver but I'm not able to test this issue. After starting a Selenium session, session id in response is nil.
```
{:ok,
 %Wallaby.Session{
   capabilities: %{
     browserName: "firefox",
     javascriptEnabled: true,
     "moz:firefoxOptions": %{
       args: ["-headless"],
       prefs: %{
         "general.useragent.override" => "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2228.0 Safari/537.36"
       }
     }
   },
   driver: Wallaby.Selenium,
   id: nil,
   screenshots: [],
   server: :none,
   session_url: "http://localhost:4444/session/",
   url: "http://localhost:4444/session/"
 }}
 ```
 
 I'm not sure why `id` is nil 🤔 
 
Also I've tried all possible advices for ChromeDriver to access to `mediaDevices` in navigator interface but I still get `nil`. I ran out of ideas :)

Do you maybe have something else on your mind what can go wrong?
